src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/GlobalFlowStateDictionaryAnalysis/GlobalFlowStateDictionaryAnalysis.cs

```diff
@@ -0,0 +1,105 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.FlowAnalysis;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis;
+
+namespace Analyzer.Utilities.FlowAnalysis.Analysis.GlobalFlowStateDictionaryAnalysis
+{
+    using GlobalFlowStateDictionaryAnalysisData = DictionaryAnalysisData<AnalysisEntity, GlobalFlowStateDictionaryAnalysisValue>;
+    using GlobalFlowStateDictionaryAnalysisResult = DataFlowAnalysisResult<GlobalFlowStateDictionaryBlockAnalysisResult, GlobalFlowStateDictionaryAnalysisValue>;
+    using GlobalFlowStateDictionaryAnalysisDomain = MapAbstractDomain<AnalysisEntity, GlobalFlowStateDictionaryAnalysisValue>;
+
+    /// <summary>
+    /// An analysis that tracks the state of a set of <see cref="AnalysisEntity"/>. The state is shared among the block.
+    /// </summary>
+    internal class GlobalFlowStateDictionaryAnalysis : ForwardDataFlowAnalysis<
+        GlobalFlowStateDictionaryAnalysisData,
+        GlobalFlowStateDictionaryAnalysisContext,
+        GlobalFlowStateDictionaryAnalysisResult,
+        GlobalFlowStateDictionaryBlockAnalysisResult,
+        GlobalFlowStateDictionaryAnalysisValue>
+    {
+        public static readonly GlobalFlowStateDictionaryAnalysisDomain Domain = new(GlobalFlowStateDictionaryAnalysisValueDomain.Instance);
+
+        public GlobalFlowStateDictionaryAnalysis(
+            AbstractAnalysisDomain<GlobalFlowStateDictionaryAnalysisData> analysisDomain,
+            DataFlowOperationVisitor<GlobalFlowStateDictionaryAnalysisData, GlobalFlowStateDictionaryAnalysisContext, GlobalFlowStateDictionaryAnalysisResult, GlobalFlowStateDictionaryAnalysisValue> operationVisitor)
+            : base(analysisDomain, operationVisitor)
+        {
+        }
+
+        public static GlobalFlowStateDictionaryAnalysisResult? TryGetOrComputeResult(
+            ControlFlowGraph cfg,
+            ISymbol owningSymbol,
+            Func<GlobalFlowStateDictionaryAnalysisContext, GlobalFlowStateDictionaryFlowOperationVisitor> createOperationVisitor,
+            WellKnownTypeProvider wellKnownTypeProvider,
+            AnalyzerOptions analyzerOptions,
+            DiagnosticDescriptor rule,
+            bool pessimisticAnalysis,
+            InterproceduralAnalysisKind interproceduralAnalysisKind = InterproceduralAnalysisKind.None,
+            InterproceduralAnalysisPredicate? interproceduralAnalysisPredicate = null)
+        {
+            if (cfg == null)
+            {
+                throw new ArgumentNullException(nameof(cfg));
+            }
+
+            var interproceduralAnalysisConfig = InterproceduralAnalysisConfiguration.Create(
+                analyzerOptions,
+                rule,
+                cfg,
+                wellKnownTypeProvider.Compilation,
+                interproceduralAnalysisKind);
+
+            var pointsToAnalysisResult = PointsToAnalysis.TryGetOrComputeResult(
```

Comment:
This may block enabling this rule by default. I agree that this looks to be algorithmically the best approach, but the perf overhead here might be too much. Using `PartialWithoutTrackingFieldsAndProperties` makes it less expensive, but we would still need extensive perf measurements to enable the rule by default.

Would it be possible to take an alternate approach by avoiding points to analysis and instead just override `SetAbstractValueForAssignment` in global flow state operation visitor and flow the global flow state values from source analysis entity to target analysis entity? That should be lot cheaper performance wise.

Reply:
By the way, it might be okay to use points to analysis for now, and tackle the perf issue later along with performance measurements, whenever we want to enable the rule by default. You may want to open a tracking issue for it though.

Reply:
How about conditional turn this on for different scenarios?
PointToAnalysis would only be useful if a local or parameter is assigned to other IEnumerable instances,
we can check if there is such operation.
And for the rest of the scenario, we can just track the AnalysisEntity directly.

Reply:
As discuss offline, I will try to test and measure the perf to help us

Reply:
I saw no big diff when run this benchmark https://github.com/dotnet/performance/blob/main/src/benchmarks/real-world/Roslyn/StageBenchmarks.cs
When the analyzer is disabled by default:
![image](https://user-images.githubusercontent.com/24360909/140868317-9084c289-0c2c-4f64-a2cc-680fc58ce1ba.png)


When the analyzer is enabled by default:

![image](https://user-images.githubusercontent.com/24360909/140865103-2315536e-91a5-4566-8908-8b8343e49caa.png)